python-package/xgboost/compat.py

```diff
@@ -55,20 +56,51 @@ def lazy_isinstance(instance: Any, module: str, name: str) -> bool:
         from sklearn.cross_validation import KFold as XGBKFold
         from sklearn.cross_validation import StratifiedKFold as XGBStratifiedKFold
 
+    # sklearn.utils Tags types can be imported unconditionally once
+    # xgboost's minimum scikit-learn version is 1.6 or higher
+    try:
+        from sklearn.utils import ClassifierTags as _sklearn_ClassifierTags
+        from sklearn.utils import RegressorTags as _sklearn_RegressorTags
+        from sklearn.utils import Tags as _sklearn_Tags
+    except ImportError:
+        _sklearn_ClassifierTags = object
+        _sklearn_RegressorTags = object
+        _sklearn_Tags = object
+
     SKLEARN_INSTALLED = True
 
 except ImportError:
     SKLEARN_INSTALLED = False
 
     # used for compatibility without sklearn
-    XGBModelBase = object
-    XGBClassifierBase = object
-    XGBRegressorBase = object
-    LabelEncoder = object
+    class XGBModelBase:  # type: ignore[no-redef]
+        """Dummy class for sklearn.base.BaseEstimator."""
+
+        pass
+
+    class XGBClassifierBase:  # type: ignore[no-redef]
+        """Dummy class for sklearn.base.ClassifierMixin."""
+
+        pass
+
+    class XGBRegressorBase:  # type: ignore[no-redef]
+        """Dummy class for sklearn.base.RegressorMixin."""
+
+        pass
+
+    class LabelEncoder:  # type: ignore[no-redef]
+        """Dummy class for sklearn.preprocessing.LabelEncoder."""
+
+        pass
```

Comment:
With all of these placeholder classes set to `object`, using the re-arranged inheritance order () results in errors like this when importing `xgboost` when `scikit-learn` is not installed:

> TypeError: Cannot create a consistent method resolution order (MRO) for bases object, XGBModel

([build link](https://github.com/dmlc/xgboost/actions/runs/12025279025/job/33522167020?pr=11021))

Having each one be a different class resolves that. I forgot until I saw this test failure that we faced a similar thing a few years ago in LightGBM: https://github.com/microsoft/LightGBM/pull/3192

Just copying @StrikerRUS 's solution from there here... it's worked well for us.